google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/Publisher.java

```diff
@@ -159,37 +151,35 @@ private Publisher(Builder builder) throws IOException {
     messagesBundle = new LinkedList<>();
     messagesBundleLock = new ReentrantLock();
     activeAlarm = new AtomicBoolean(false);
-    int numCores = Math.max(1, Runtime.getRuntime().availableProcessors());
     executor = builder.executorProvider.getExecutor();
     if (builder.executorProvider.shouldAutoClose()) {
       closeables.add(
           new AutoCloseable() {
             @Override
-            public void close() throws IOException {
+            public void close() {
               executor.shutdown();
             }
           });
     }
-    channels = new Channel[numCores];
-    channelIndex = new AtomicRoundRobin(channels.length);
-    for (int i = 0; i < numCores; i++) {
+    channels = new ManagedChannel[Runtime.getRuntime().availableProcessors()];
+    for (int i = 0; i < channels.length; i++) {
       channels[i] =
-          builder.channelBuilder.isPresent()
-              ? builder.channelBuilder.get().build()
-              : NettyChannelBuilder.forAddress(
-                      PublisherSettings.getDefaultServiceAddress(),
-                      PublisherSettings.getDefaultServicePort())
-                  .negotiationType(NegotiationType.TLS)
-                  .sslContext(GrpcSslContexts.forClient().ciphers(null).build())
-                  .executor(executor)
-                  .build();
+          builder.channelProvider.needsExecutor()
+              ? builder.channelProvider.getChannel(executor)
+              : builder.channelProvider.getChannel();
```

Comment:
Hmm after offline discussion, it looks like only accepting `InstantiatingChannelProvider` might not make sense; it makes testing harder. We should just be clear in the documentation.